examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,635 @@
+"""
+Title: Natural Language Search with Dual Encoder
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/30
+Last modified: 2021/01/30
+Description: Implementing a dual encoder model for image search with natural language.
+"""
+
+"""
+## Introduction
+
+The example demonstrates how to build a dual encoder (also known as two-tower) neural network
+model to search for images using natural language. The model is inspired by
+the [CLIP](https://openai.com/blog/clip/)
+approach, introduced by Alec Radford et al. The idea is to train a vision encoder and a text
+encoder jointly to project the representation of images and their captions into the same embedding
+space, such that the caption embeddings are located near the embeddings of the images they describe.
+
+This example requires TensorFlow 2.4 or higher.
+In addition, [TensorFlow Hub](https://www.tensorflow.org/hub)
+and [TensorFlow Text](https://www.tensorflow.org/tutorials/tensorflow_text/intro)
+are required for the BERT model, and [TensorFlow Addons](https://www.tensorflow.org/addons)
+is required for the AdamW optimizer. These libraries can be installed using the
+following command:
+
+```python
+pip install -q -U tensorflow-hub tensorflow-text tensorflow-addons
+```
+"""
+
+"""
+## Setup
+"""
+
+import os
+import collections
+import json
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_hub as hub
+import tensorflow_text as text
+import tensorflow_addons as tfa
+import matplotlib.pyplot as plt
+import matplotlib.image as mpimg
+from tqdm import tqdm
+
+# Suppressing tf.hub warnings
+tf.get_logger().setLevel("ERROR")
+
+"""
+## Prepare the data
+
+We will use the [MS-COCO](https://cocodataset.org/#home) dataset to train our
+dual encoder model. It contains over 82,000 images, each of which has at least
```

Comment:
Suggestion:

"It" may appear to be referring to the "model", because of the order of the words in the previous sentence, so:

```suggestion
dual encoder model. MS-COCO contains over 82,000 images, each of which has at least
```